File: 프로그래머스/lv2/17684. ［3차］ 압축/［3차］ 압축.py
def solution(msg):
    answer = []
    idx ={}
    for i,v in enumerate("ABCDEFGHIJKLMNOPQRSTUVWXYZ",1):
        idx[v]=i
    s=""
    cur=0
    msg+=" "
    while cur<len(msg):
        s+=msg[cur]
        if s==" ":
            break
        if s not in idx:
            idx[s]=len(idx)+1
            answer.append(idx[s[:-1]])
            s=""
            continue
        cur+=1
    return answer

File: 프로그래머스/2/17684. ［3차］ 압축/［3차］ 압축.java
import java.util.*;

class Solution {
    public int[] solution(String msg) {
        Map<String, Integer> map = new HashMap<>();
        for(int i=1; i<=26; i++){
            map.put(Character.toString('A'+i-1), i);
        }
        
        List<Integer> list = new ArrayList<>();
        
        ArrayDeque<String> q = new ArrayDeque<>();
        for(int i=0; i<msg.length(); i++){
            q.offer(Character.toString(msg.charAt(i)));
        }
        StringBuilder sb = new StringBuilder();
        while(!q.isEmpty()){
            String tmp = q.poll();
            sb.append(tmp);
            if(!map.containsKey(sb.toString())){
                map.put(sb.toString(), map.size()+1);
                sb.deleteCharAt(sb.length()-1).toString();
                list.add(map.get(sb.toString()));
                q.offerFirst(tmp);
                sb.setLength(0);
            }
        }
        list.add(map.get(sb.toString()));
        
        int[] answer = new int[list.size()];
        for(int i=0; i<list.size(); i++){
            answer[i] = list.get(i);
        }
        return answer;
    }
}